llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp

```diff
@@ -7638,7 +7644,7 @@ LegalizerHelper::lowerSADDO_SSUBO(MachineInstr &MI) {
   // (LHS) if and only if the other operand (RHS) is (non-zero) positive,
   // otherwise there will be overflow.
   auto ResultLowerThanLHS =
-      MIRBuilder.buildICmp(CmpInst::ICMP_SLT, BoolTy, Dst0, LHS);
+      MIRBuilder.buildICmp(CmpInst::ICMP_SLT, BoolTy, Tmp, LHS);
```

Comment:
@aemerson Like in this way?

Reply:
Does it also mean that it is generally not recommended to write like `Builder.buildInstr(MI.getOperand(0).getReg(), ...);`?

Reply:
I didn't realize this issue was quite that prevalent. We don't have an explicit contract with what state the MIR is expected to be in when we use MIRBuilder, and maybe we should.

`Builder.buildInstr(MI.getOperand(0).getReg(), ...);` only works if `buildInstr` is a dumb constructor, but for CSEMIRBuilder that's not true. I'm not totally sure what the right thing to do is. @arsenm any opinions?

Perhaps we should be strict and say that you can't construct another def of a vreg, so `MI.eraseFromParent()` needs to be called before you do so.

Alternatively, maybe CSEMachineIRBuilder can automatically erase the original instruction if any of the `DstOp` operands are `Register`. These should be separate changes from this patch though.

Reply:
Last time I looked at the CSEMIRBuilder I couldn't decide what to do. If you pass in existing registers, you lose out on the auto-folding, but it's not broken

Reply:
@shiltian  What if we assert in `buildInstr()` that `DstOp` vregs are undefined? That will then force clients to `eraseFromParent()` *before* calling it.

The other idea of having `buildInstr()` erase the MI itself forces each vreg build to call `MRI.getVRegDef()` even if it's never needed, and it's probably unintuitive from an API perspective.

Reply:
Yeah, I agree that assertion sounds better. We definitely don't want to have any hidden behavior in those interfaces. If a MI is erased, this will "reset" the destination register back to a generic one right, since the definition is gone?